dc/s2n-quic-dc/src/path/secret/receiver.rs

```diff
@@ -301,70 +62,73 @@ pub enum Error {
+        // TODO: Provide more useful pre-auth checks. For now just don't bother checking this, we
+        // can always rely on the post-auth check in practice, this is just a slight optimization.
         Ok(())
     }
 
     pub fn minimum_unseen_key_id(&self) -> KeyId {
-        KeyId::try_from(self.max_seen_key_id.load(Ordering::Relaxed) + 1).unwrap()
+        KeyId::try_from(
+            self.max_seen_key_id
+                .load(Ordering::Relaxed)
+                // Initial u64::MAX wraps to zero, which is the correct answer for the initial
+                // state. After that just +1 consistently.
+                .wrapping_add(1),
+        )
+        .unwrap()
     }
 
     /// Called after decryption has been performed
     pub fn post_authentication(&self, identity: &Credentials) -> Result<(), Error> {
-        let key_id = identity.key_id;
-        self.max_seen_key_id.fetch_max(*key_id, Ordering::Relaxed);
-        let pn = PacketNumberSpace::Initial.new_packet_number(key_id);
-
-        // Note: intentionally retaining this lock across potential insertion into the shared map.
-        // This avoids the case where we have evicted an entry but cannot see it in the shared map
-        // yet from a concurrent thread. This should not be required for correctness but helps
-        // reasoning about the state of the world.
         let mut seen = self.seen.lock().unwrap();
-        match seen.insert_with_evicted(pn) {
-            Ok(evicted) => {
-                if let Some(shared) = &self.shared {
-                    // FIXME: Consider bounding the number of evicted entries to insert or
-                    // otherwise optimizing? This can run for at most 128 entries today...
-                    for evicted in evicted {
-                        shared.insert(&Credentials {
-                            id: identity.id,
-                            key_id: PacketNumber::as_varint(evicted),
-                        });
-                    }
-                }
-                Ok(())
-            }
-            Err(SlidingWindowError::TooOld) => {
-                if let Some(shared) = &self.shared {
-                    shared.remove(identity)
-                } else {
-                    Err(Error::Unknown)
-                }
-            }
-            Err(SlidingWindowError::Duplicate) => Err(Error::AlreadyExists),
+
+        let key_id = *identity.key_id;
+        let mut previous_max = self.max_seen_key_id.load(Ordering::Relaxed);
+        let new_max = if previous_max == u64::MAX {
+            previous_max = 0;
+            key_id
+        } else {
+            previous_max.max(key_id)
+        };
+        self.max_seen_key_id.store(new_max, Ordering::Relaxed);
+
+        let delta = new_max - previous_max;
+        if delta > seen.len() as u64 {
+            // not yet seen since we shifted forward by more than the bitset's size.
+            seen.fill(false);
```

Comment:
nit: It looks like the bitvec impl already does this optimization: https://docs.rs/bitvec/latest/src/bitvec/slice.rs.html#1481

Reply:
That's ==, this is > where bitvec treats it as a panic (see just below the line you linked).

Reply:
ah right. yours is better :D